src/gui/qgsfieldmappingmodel.cpp

```diff
@@ -0,0 +1,434 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsfieldmappingmodel.h"
+#include "qgsexpressioncontextutils.h"
+
+
+QgsFieldMappingModel::QgsFieldMappingModel( const QgsFields &sourceFields,
+    const QgsFields &destinationFields,
+    const QMap<QString, QgsExpression> &expressions,
+    QObject *parent )
+  : QAbstractTableModel( parent )
+  , mSourceFields( sourceFields )
+  , mExpressionContextGenerator( new ExpressionContextGenerator( &mSourceFields ) )
+{
+  setDestinationFields( destinationFields, expressions );
+}
+
+QVariant QgsFieldMappingModel::headerData( int section, Qt::Orientation orientation, int role ) const
+{
+  if ( role == Qt::DisplayRole )
+  {
+    if ( orientation == Qt::Horizontal )
+    {
+      switch ( section )
+      {
+        case static_cast<int>( ColumnDataIndex::SourceExpression ):
+        {
+          return tr( "Source expression" );
+        }
+        case static_cast<int>( ColumnDataIndex::DestinationName ):
+        {
+          return tr( "Name" );
+        }
+        case static_cast<int>( ColumnDataIndex::DestinationType ):
+        {
+          return tr( "Type" );
+        }
+        case static_cast<int>( ColumnDataIndex::DestinationLength ):
+        {
+          return tr( "Length" );
+        }
+        case static_cast<int>( ColumnDataIndex::DestinationPrecision ):
+        {
+          return tr( "Precision" );
+        }
+        case static_cast<int>( ColumnDataIndex::DestinationConstraints ):
+        {
+          return tr( "Constraints" );
+        }
+      }
+    }
+    else if ( orientation == Qt::Vertical )
+    {
+      return section;
+    }
+  }
+  return QVariant();
+}
+
+QgsFields QgsFieldMappingModel::sourceFields() const
+{
+  return mSourceFields;
+}
+
+int QgsFieldMappingModel::rowCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent );
+  return mMapping.count();
+}
+
+int QgsFieldMappingModel::columnCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent );
+  return 6;
+}
+
+QVariant QgsFieldMappingModel::data( const QModelIndex &index, int role ) const
+{
+  if ( index.isValid() )
+  {
+    const int col { index.column() };
+    const Field &f { mMapping.at( index.row() ) };
+
+    const QgsFieldConstraints::Constraints constraints { fieldConstraints( f.field ) };
+
+    if ( role == Qt::DisplayRole || role == Qt::EditRole )
+    {
+      switch ( col )
+      {
+          {
+          case static_cast<int>( ColumnDataIndex::SourceExpression ):
```

Comment:
are these static_casts actually needed? Might be nicer to static_cast `col` rather then these values

Reply:
they are needed after the `enum` became a `class enum` but I agree that it is better to case the `col`